ts/indicators/regressions.src.ts

```diff
@@ -0,0 +1,601 @@
+         * @param {Array<number>} xData -  list of all x coordinates in a period
+         * @param {Array<number>} yData - list of all y coordinates in a period
+         * @return {Highcharts.RegressionLineParametersObject}
+         *          object that contains the slope and the intercept
+         *          of a straight line function
+         */
+        getRegressionLineParameters: function (
+            this: Highcharts.LinearRegressionIndicator,
+            xData: Array<number>,
+            yData: Array<number>
+        ): Highcharts.RegressionLineParametersObject {
+            // least squares method
+            var yIndex: number = (this.options.params as any).index,
+                getSingleYValue = function (
+                    yValue: (number|Array<number>),
+                    yIndex: number
+                ): number {
+                    return isArray(yValue) ? yValue[yIndex] : yValue;
+                },
+                xSum = xData.reduce(
+                    function (
+                        accX: number,
+                        val: number
+                    ): number {
+                        return val + accX;
+                    }, 0),
+                ySum = yData.reduce(
+                    function (accY: number, val: number): number {
+                        return getSingleYValue(val, yIndex) + accY;
+                    }, 0),
+                xMean: number = xSum / xData.length,
+                yMean: number = ySum / yData.length,
+                xError: number,
+                yError: number,
+                formulaNumerator = 0,
+                formulaDenominator = 0,
+                i: number,
+                slope: number;
+
+            for (i = 0; i < xData.length; i++) {
+                xError = xData[i] - xMean;
+                yError = getSingleYValue(yData[i], yIndex) - yMean;
+                formulaNumerator += xError * yError;
+                formulaDenominator += Math.pow(xError, 2);
+            }
+
+            slope = formulaDenominator ?
+                formulaNumerator / formulaDenominator : 0; // don't divide by 0
+
+            return {
+                slope: slope,
+                intercept: yMean - slope * xMean
+            };
+        },
+
+
+        /**
+         * Return the y value on a straight line.
+         * @private
+         * @param {Highcharts.RegressionLineParametersObject} lineParameters
+         *          object that contains the slope and the intercept
+         *          of a straight line function
+         * @param {number} endPointX - x coordinate of the point
+         * @return {number} - y value of the point that lies on the line
+         */
+        getEndPointY: function (
+            lineParameters: Highcharts.RegressionLineParametersObject,
+            endPointX: number
+        ): number {
+            return (lineParameters.slope as any) * endPointX +
+                (lineParameters.intercept as any);
```

Comment:
The interface should have the properties non-optional, so that you can remove the `any` cast.

Reply:
While I wholeheartedly agree it is a bit unclear to me as to when it is ok to make interfaces with non-optional properties in phase one of the migration. If it is used as a pure value object it is ok?